### PR TITLE
Implemented GLEM for node-level tasks

### PR DESCRIPTION
**Pull request by wangz10**
*Issue #, if available:* #241

*Description of changes:* Implemented GLEM for node-level tasks. It's implemented by subclassing `GSgnnNodeModelBase` for the model, and a specialized trainer `GLEMNodePredictionTrainer` for alternating training of LM and GNN.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by wangz10** [comment]
@zheng-da @classicsong , Can you check the error log in [e2e_check](https://github.com/awslabs/graphstorm/actions/runs/5322481594/jobs/9639415588?pr=262)? Is it a time-out error? I didn't add any new e2e test. My new test in e2e_mgpu passed.

File: python/graphstorm/gsf.py [hunk 1]
```diff
@@ -150,7 +151,10 @@ def create_builtin_node_model(g, config, train_task):
     -------
     GSgnnModel : The GNN model.
     """
-    model = GSgnnNodeModel(config.alpha_l2norm)
+    if hasattr(config, "_glem"):
```
Comment (zheng-da) [inline python/graphstorm/gsf.py:154]: you should create a `glem` property in config and just use `config.glem`.

File: python/graphstorm/model/node_glem.py [hunk 2]
```diff
@@ -0,0 +1,267 @@
...
+
+class GLEM(GSgnnNodeModelBase):
+    """
+    GLEM model (https://arxiv.org/abs/2210.14709) for node-level tasks.
+    """
+    def __init__(self,
+                 alpha_l2norm,
+                 em_order_gnn_first=True,
+                 inference_using_gnn=True,
+                 pl_weight=0.5
```
Comment (zheng-da) [inline python/graphstorm/model/node_glem.py:33]: explain the input arguments.

File: python/graphstorm/model/node_glem.py [hunk 3]
```diff
@@ -0,0 +1,267 @@
...
+        self.alpha_l2norm = alpha_l2norm
+        self.em_order_gnn_first = em_order_gnn_first
+        self.inference_using_gnn = inference_using_gnn
+        self.pl_weight = pl_weight
+        self.lm = GSgnnNodeModel(alpha_l2norm)
+        self.gnn = GSgnnNodeModel(alpha_l2norm)
+        self.training_lm = not em_order_gnn_first
+        # self.lm.node_input_encoder: graphstorm.model.lm_embed.GSLMNodeEncoderInputLayer
+        # self.lm.gnn_encoder: None
+        # self.lm.decoder: EntityClassifier
```
Comment (zheng-da) [inline python/graphstorm/model/node_glem.py:45]: please remove the commented code.

File: python/graphstorm/model/node_glem.py [hunk 4]
```diff
@@ -0,0 +1,267 @@
...
+    def toggle(self, part='lm'):
+        """The method toggles training between lm and gnn."""
+        if part == 'lm':
+            self.training_lm = True
+            self.freeze_params('gnn')
+            self.unfreeze_params('lm')
+        elif part == 'gnn':
+            self.training_lm = False
+            self.freeze_params('lm')
+            self.unfreeze_params('gnn')
```
Comment (zheng-da) [inline python/graphstorm/model/node_glem.py:148]: you should check the value of the input argument.

File: python/graphstorm/model/node_glem.py [hunk 5]
```diff
@@ -0,0 +1,267 @@
...
+        # L2 regularization of dense parameters
+        for d_para in self.gnn.get_dense_params():
+            reg_loss += d_para.square().sum()
+        # weighted addition to the total loss
+        return loss + self.alpha_l2norm * reg_loss
+
+
+    def forward_lm(self, blocks, node_feats, _, labels, input_nodes):
+        """Forward pass for node prediction using LM"""
+        emb = self._embed_nodes(blocks, node_feats, _, input_nodes)
```
Comment (zheng-da) [inline python/graphstorm/model/node_glem.py:196]: In GLEM, we also need to compute GNN for both E and M stages? if so, it'll be very expensive to fine-tune the BERT model

**Comment by zheng-da** [comment]
did you evaluate the model performance of this implementation on some public datasets?

File: python/graphstorm/model/node_glem.py [hunk 6]
```diff
@@ -0,0 +1,267 @@
...
+        labels = self._process_labels(labels)
+        logits = self.lm.decoder(emb)
```
Comment (zheng-da) [inline python/graphstorm/model/node_glem.py:198]: this doesn't look right. the embeddings here are GNN embeddings.

File: python/graphstorm/model/node_glem.py (hunk 5 above)
Comment (wangz10) [reply python/graphstorm/model/node_glem.py:196]: Yes. In E stage (LM fine-tuning), need to run GNN to generate pseudo labels for supervising LM.

File: python/graphstorm/config/argument.py [hunk 7]
```diff
@@ -386,6 +393,21 @@ def freeze_lm_encoder_epochs(self):
 
         return 0
 
+    @property
+    def glem(self):
+        """ Setting up (default) GLEM configs 
+        """
+        if hasattr(self, "_glem"):
```
Comment (zheng-da) [inline python/graphstorm/config/argument.py:400]: i don't understand how it is set up. users only need to turn on the flag?

File: python/graphstorm/model/node_glem.py [hunk 8]
```diff
@@ -0,0 +1,311 @@
...
+                                         lr=lm_lr if lm_lr is not None else lr,
+                                         weight_decay=weight_decay)
+            lm_opts = [lm_optimizer]
+        else:
+            lm_opts = []
+        self._optimizer = GSOptimizer(dense_opts=dense_opts,
+                                      lm_opts=lm_opts,
+                                      sparse_opts=sparse_opts)
+
+    def create_optimizer(self):
```
Comment (zheng-da) [inline python/graphstorm/model/node_glem.py:85]: is the code covered by lint? it requires a docstring.

File: python/graphstorm/model/node_glem.py [hunk 9]
```diff
@@ -0,0 +1,311 @@
...
+        """Set the same decoder for both, since lm needs to be able to 
+        predict node labels during training
+        """
+        self.lm.set_decoder(deepcopy(decoder))
+        self.gnn.set_decoder(decoder)
+
+    def set_loss_func(self, loss_fn):
+        """Set the final loss function based on the node task 
+        """
+        self.lm.set_loss_func(loss_fn)
```
Comment (zheng-da) [inline python/graphstorm/model/node_glem.py:125]: is it possible that the loss function has parameters?

File: python/graphstorm/model/node_glem.py [hunk 10]
```diff
@@ -0,0 +1,311 @@
...
+        # add regularization loss to all parameters to avoid the unused parameter errors
+        reg_loss = th.tensor(0.).to(loss.device)
+        # L2 regularization of dense parameters
+        for d_para in self.gnn.get_dense_params():
+            reg_loss += d_para.square().sum()
+        # weighted addition to the total loss
+        return loss + self.alpha_l2norm * reg_loss
+
+
+    def predict(self, blocks, node_feats, edge_feats, input_nodes, return_proba):
```
Comment (zheng-da) [inline python/graphstorm/model/node_glem.py:241]: this one doesn't have docstring either.

File: python/graphstorm/config/argument.py (hunk 7 above)
Comment (wangz10) [reply python/graphstorm/config/argument.py:400]: Yes. Please see the example config file: [training_scripts/gsgnn_np/ml_nc_utext_glem.yml](https://github.com/awslabs/graphstorm/pull/262/files/7ef2d3751671094e48aeff17e7130dbbb5f2c9ae#diff-c86b86ea495b966bb658770fdefb4414d74bea3242e0718167fc4b5de397b036).

File: python/graphstorm/model/node_glem.py (hunk 8 above)
Comment (wangz10) [reply python/graphstorm/model/node_glem.py:85]: Not covered by lint somehow, will add.

File: python/graphstorm/model/node_glem.py (hunk 9 above)
Comment (wangz10) [reply python/graphstorm/model/node_glem.py:125]: That might be possible. However, the `set_loss_func` calls `GSgnnModle.set_loss_func`, which requires `loss_fn` to be a `nn.Module`. Parameters should be passed when initialize the `loss_fn` Module object.

File: python/graphstorm/model/node_glem.py (hunk 10 above)
Comment (wangz10) [reply python/graphstorm/model/node_glem.py:241]: Will add

File: python/graphstorm/model/node_glem.py (hunk 9 above)
Comment (zheng-da) [reply python/graphstorm/model/node_glem.py:126]: if `loss_fn` contains parameters, do we want the loss functions in LM and GNN shared?

File: python/graphstorm/model/node_glem.py (hunk 8 above)
Comment (zheng-da) [reply python/graphstorm/model/node_glem.py:85]: can you add lint check?

File: python/graphstorm/model/node_glem.py (hunk 9 above)
Comment (wangz10) [reply python/graphstorm/model/node_glem.py:126]: Yes. Their `loss_fn` should be identical. This is the supervised loss (cross-entropy) for labeled nodes in GLEM.

File: python/graphstorm/model/node_glem.py (hunk 8 above)
Comment (wangz10) [reply python/graphstorm/model/node_glem.py:85]: I tried to play around adding it but it conflicts with `@abstractmethod` and its docstring defined in parent class. This seems to be a known issue in pylint: https://github.com/pylint-dev/pylint/issues/4916

File: python/graphstorm/model/node_glem.py [hunk 11]
```diff
@@ -0,0 +1,336 @@
...
+        emb_lm, emb_gnn = self._embed_nodes(blocks, node_feats, _, input_nodes, do_gnn_encode=True)
+        labels = self._process_labels(labels)
+        logits = self.gnn.decoder(emb_gnn)
+        if no_pl:
+            loss = self.gnn.loss_func(logits, labels)
+        else:
+            batch_size = labels.size(0)
+            n_gold_nodes = batch_size // 2
+            # compute pseudo labels from LM
+            logits_lm = self.lm.decoder(emb_lm[n_gold_nodes:])
```
Comment (zheng-da) [inline python/graphstorm/model/node_glem.py:198]: this PR uses the existing training set to train the GNN model with LM or vice versa. do we expect to get better performance if we take advantage of the unlabelled nodes? does the original paper only use the training set to train the model?

File: training_scripts/gsgnn_np/ml_nc_utext_glem.yml [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+    part_config: movielen_100k_utext_train_val_1p_4t/movie-lens-100k.json
+    verbose: false
+    save_perf_results_path: null
+  lm:
+    lm_train_nodes: 10
+    lm_infer_batch_size: 32
+    freeze_lm_encoder_epochs: 0
+    glem:
+      em_order_gnn_first: true
+      inference_using_gnn: true
```
Comment (zheng-da) [inline training_scripts/gsgnn_np/ml_nc_utext_glem.yml:25]: is this argument necessary?

File: training_scripts/gsgnn_np/ml_nc_utext_glem.yml [hunk 13]
```diff
@@ -0,0 +1,56 @@
...
+    backend: gloo
+    ip_config: ip_list.txt
+    part_config: movielen_100k_utext_train_val_1p_4t/movie-lens-100k.json
+    verbose: false
+    save_perf_results_path: null
+  lm:
+    lm_train_nodes: 10
+    lm_infer_batch_size: 32
+    freeze_lm_encoder_epochs: 0
+    glem:
```
Comment (zheng-da) [inline training_scripts/gsgnn_np/ml_nc_utext_glem.yml:23]: should we regard this as a training method?

**Comment by zheng-da** [comment]
overall, it looks good to me. a few overall questions:
* can this reproduce the performance in the paper?
* should we use unlabelled nodes to train the model as well?

File: python/graphstorm/model/node_glem.py (hunk 11 above)
Comment (wangz10) [reply python/graphstorm/model/node_glem.py:198]: We expect to get better performance if we use unlabeled nodes. The original paper uses unlabeled nodes in transductive setting. They also have experiments in inductive setting.

File: training_scripts/gsgnn_np/ml_nc_utext_glem.yml (hunk 12 above)
Comment (wangz10) [reply training_scripts/gsgnn_np/ml_nc_utext_glem.yml:25]: Yes. I didn't implement `em_order_gnn_first`. Will add in the next commit.

File: training_scripts/gsgnn_np/ml_nc_utext_glem.yml (hunk 13 above)
Comment (wangz10) [reply training_scripts/gsgnn_np/ml_nc_utext_glem.yml:23]: yes

**Review by zheng-da — APPROVED** [review]
